### DISCUSSION TITLE
Using pseudo instructions to invoke a method of a simobject

### DISCUSSION THREAD

**Discussion by kaplannp** [gem5/gem5#1348] [Q&A] [answered]
Hi all. Gem5 beginner here. I'm trying to hack the simulator to add a new SimObject and invoke one of its methods through a pseudo instruction.

I've registered my pseudo instruction in these files
```
include/gem5/m5ops.h
include/gem5/asm/generic/m5ops.h
util/m5/src/command/[INSTRUCTION].cc 
```

Then I add the line `tc->getSystemPtr()->MySimObject->myMethod()` to a new pseudo instruction in `src/sim/pseudo_inst.cc`, but when I go to compile I get the error
`src/sim/pseudo_inst.cc:568:30: error: ‘class gem5::System’ has no member named MySimObject`

Is this the right way to go about this? Where can I instantiate `MySimObject` so it will be available to me in the pseudo instruction?

**Comment by robhau** [comment, marked as answer]
Hi @kaplannp,

I would propose you another design to to achieve your desired behavior.

Usually, you can configure peripherals (I assume that your SimObject is a peripheral) by writing to memory mapped registers. They are also used to trigger specific tasks. So I suggest that you implement register(s) in your SimObject that corresponds to your task(s). When write, for example, a 1 to a task register, you then call your member function. 

Gem5 already offers you register banks that you can use to implement registers (see `src/dev/reg_bank.hh`). Take a look in SimObjects where they are used to learn how they work. For example, they are part of `Plic` and `Clint` (`src/dev/riscv/*`)
